### PR TITLE
Fixing issue #4112 Replace deprecated SHILCreateFromPath with SHParseDisplayName

### PR DESCRIPTION
**Pull request by wreuel**
…DisplayName

Fixes #

Issue #4112 changed the deprecated as should be

## Proposed changes

- Rewrite the code to use the SHParseDisplayName 

## Customer Impact

- None

## Regression? 

-   No

## Risk

- No

## Test methodology 

- Run the WinformsControlsTest
- Open the Dialogs Samples
- Breakpoint on the lines where GetShellItemForPath 
- Checked the behaviour when the paths is provided and when is not provided

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/4707)

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/winforms?pullRequest=4707) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/winforms?pullRequest=4707) before we can accept your contribution.<br/><br/>:x: wreuel [sign now](https://cla.dotnetfoundation.org/dotnet/winforms?pullRequest=4707)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/winforms?pullRequest=4707) it.</sub>

**Comment by wreuel** [comment]
@RussKie see, it failed on the same test that are failing if I just clone the repo and run the  _.\build -test_ command, should I do something else? Hope we can solve this.

File: src/System.Windows.Forms.Primitives/src/Interop/Shell32/Interop.SHCreateShellItem.cs [hunk 1]
```diff
@@ -16,7 +16,7 @@ internal static partial class Shell32
         public static IShellItem GetShellItemForPath(string path)
         {
             uint zero = 0;
-            if (SHILCreateFromPath(path, out IntPtr pidl, ref zero).Succeeded())
+            if (SHParseDisplayName(path, IntPtr.Zero, out IntPtr pidl, 0, out zero).Succeeded())
```
Comment (RussKie) [inline src/System.Windows.Forms.Primitives/src/Interop/Shell32/Interop.SHCreateShellItem.cs:19]: If you're doing it as an `out` param, you can do this:
```suggestion
            if (SHParseDisplayName(path, IntPtr.Zero, out IntPtr pidl, 0, out uint _).Succeeded())
```

**Review by RussKie — CHANGES_REQUESTED** [review]
Can you please describe how you tested the change?

**Comment by RussKie** [comment]
> run the _.\build -test_ command, should I do something else? Hope we can solve this.

Depends on what tests failed. Locale specific tests are known to fail (e.g. https://github.com/dotnet/winforms/issues/2734). So as clipboard-related tests.
If you provide more info, we can look into it.

**Comment by wreuel** [comment]
> Can you please describe how you tested the change?

@RussKie  About the tests, I just used the one that existed before, that was using the SHILCreateFromPath, the test is ShellItemTests.SHParseDisplayName_ValidPath() and it is passing, if I did wrong please LMK, trying to help. :)

**Comment by wreuel** [comment]
> > run the _.\build -test_ command, should I do something else? Hope we can solve this.
> 
> Depends on what tests failed. Locale specific tests are known to fail (e.g. #2734). So as clipboard-related tests.
> If you provide more info, we can look into it.

@RussKie  I added the US-EN location to my computer, now some tests are being skipped. As you can see in the image below.


image

**Comment by RussKie** [comment]
> added the US-EN location to my computer, now some tests are being skipped. As you can see in the image below.

You're running tests in x64, and these are x86 specific tests. This is expected.

**Comment by RussKie** [comment]
> > Can you please describe how you tested the change?
> 
> About the tests, I just used the one that existed before, that was using the SHILCreateFromPath, the test is ShellItemTests.SHParseDisplayName_ValidPath() and it is passing, if I did wrong please LMK, trying to help. :)

Our automated tests don't have an exhaustive coverage, so likely there are code paths that aren't covered by those.
You also need to do an exploratory testing, i.e. run `WinformsControlsTest` test app (`Dialogs` sample) to verify that all code paths are executed, and the behaviour is expected.

`SHParseDisplayName` is only invoked through `GetShellItemForPath` which is used in three places:
image

So the bare minimum you need to run `FolderBrowserDialog`, `OpenFileDialog` and `SaveFileDialog` samples to confirm they still work as expected.

**Review by RussKie — COMMENTED** [review]
👍

**Comment by wreuel** [comment]
@RussKie  so, I did the exploratory test, as you explained. And the Dialog are working fine.
Every path that goes to the GetShellItemForPath was hit correctly and the dialog opened on the path that was provided before hand on the config options. 

Also it worked fine when the path wasn't provided.

Do you need anything else?

**Review by wreuel — COMMENTED** [review]
I have done in a new commit

**Review by RussKie — APPROVED** [review]

**Comment by RussKie** [comment]
Thank you!

**Pull request merged**